perf-client.go

```diff
@@ -0,0 +1,133 @@
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+//
+
+package madmin
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/dustin/go-humanize"
+)
+
+// ClientPerfExtraTime - time for get lock or other
+type ClientPerfExtraTime struct {
+	Dur int64 `json:"dur,omitempty"`
+}
+
+// ClientPerfResult - stats from  client to server
+type ClientPerfResult struct {
+	TX       uint64 `json:"tx,omitempty"`
+	Endpoint string `json:"endpoint,omitempty"`
+	Error    string `json:"error,omitempty"`
+}
+
+// clientPerfReader - wrap the reader
+type clientPerfReader struct {
+	count     uint64
+	startTime time.Time
+	endTime   time.Time
+	buf       []byte
+}
+
+// Start - reader start
+func (c *clientPerfReader) Start() {
+	buf := make([]byte, 128*humanize.KiByte)
+	rand.Read(buf)
+	c.buf = buf
+	c.startTime = time.Now()
+}
+
+// End - reader end
+func (c *clientPerfReader) End() {
+	c.endTime = time.Now()
+}
+
+// Read - reader send data
+func (c *clientPerfReader) Read(p []byte) (n int, err error) {
+	n = copy(p, c.buf)
+	c.count += uint64(n)
+	return n, nil
+}
+
+var _ io.Reader = &clientPerfReader{}
+
+const (
+	// MaxClientPerfTimeout for max time out for client perf
+	MaxClientPerfTimeout = time.Second * 30
+	// MinClientPerfTimeout for min time out for client perf
+	MinClientPerfTimeout = time.Second * 5
+)
+
+// ClientPerf - perform net from client to MinIO servers
+func (adm *AdminClient) ClientPerf(ctx context.Context, dur time.Duration) (result ClientPerfResult, err error) {
+	if dur > MaxClientPerfTimeout {
+		dur = MaxClientPerfTimeout
+	}
+	if dur < MinClientPerfTimeout {
+		dur = MinClientPerfTimeout
+	}
+	ctx, cancel := context.WithTimeout(ctx, dur)
+	defer cancel()
+	result.Endpoint = adm.endpointURL.String()
+	queryVals := make(url.Values)
+	reader := &clientPerfReader{}
+	reader.Start()
+	_, err = adm.executeMethod(ctx, http.MethodPost, requestData{
+		queryValues:   queryVals,
+		relPath:       adminAPIPrefix + "/speedtest/client/devnull",
+		contentReader: reader,
+	})
+	reader.End()
+	if errors.Is(err, context.DeadlineExceeded) {
+		err = nil
+	}
+
+	resp, err := adm.executeMethod(context.Background(), http.MethodPost, requestData{
+		queryValues: queryVals,
+		relPath:     adminAPIPrefix + "/speedtest/client/devnull/extratime",
+	})
+	if err != nil {
+		return ClientPerfResult{}, err
+	}
+	var extraTime ClientPerfExtraTime
+	dec := json.NewDecoder(resp.Body)
+	err = dec.Decode(&extraTime)
+	if err != nil {
+		return ClientPerfResult{}, err
+	}
+	durSpend := reader.endTime.Sub(reader.startTime).Nanoseconds()
+	if extraTime.Dur > 0 {
+		durSpend = durSpend - extraTime.Dur
+	}
+	if durSpend <= 0 {
+		return ClientPerfResult{}, fmt.Errorf("DurSpend is unexpected! Spent too much at get the lock!")
+	}
+	return ClientPerfResult{
+		TX:    uint64(float64(reader.count) / time.Duration(durSpend).Seconds()),
```

Comment:
```suggestion
		TX:    uint64(float64(reader.count) / time.Duration(durSpend).Seconds()),
```

Let's actually provide lower values from an API point of view such that `mc` can do humanize or pretty calculations.

```
ClientPerfResult{
   BytesRead: uint64(reader.count),
   TimeSpent: durSpend,
}
```